### ISSUE TITLE
PVC is always removed on every reconcilation without a change and with prune, force annotations set to disabled

### ISSUE THREAD

**Issue by kevinvalk** [fluxcd/flux2#4392]
### Describe the bug

I have a PVC like mentioned below and on the first reconcilation (check) the pvc is simply deleted again. It does not show anything related to this in the event log in the kustomization resource nor in the flux logs (even after enabling debug on everything).  (NOTE: the flux kustomization that triggers this has `prune` and `force` set to true)
```yaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: repository-pvc
spec:
  accessModes:
    - ReadWriteMany
  resources:
    requests:
      storage: 10Gi
  storageClassName: longhorn
```

I also have a hard time to reproduce this as it is currently super consistent (each and every reconcile) but as soon as you make a change (for example, add another completely separate PVC) I cannot get it to work again.

But more importantly, flux still does this even when I add the prune and force disabled annotations.
```yaml
metadata:
  name: repository-pvc
  annotations:
    kustomize.toolkit.fluxcd.io/prune: disabled
    kustomize.toolkit.fluxcd.io/force: disabled
```

_This particular pvc and usage I had in production for a bit over a year now and I have observed this behavior after switching to flux 2.1, so this may be a regression. Will check if downgrading resolves the issue._

### Steps to reproduce

This sadly is hard to reproduce....

### Expected behavior

I expect the pvc not to be deleted.

### OS / Distro

N/A

### Flux version

v2.1.2

### Flux check

flux --kubeconfig /etc/rancher/k3s/k3s.yaml check
► checking prerequisites
✔ Kubernetes 1.28.3+k3s1 >=1.25.0-0
► checking controllers
✔ helm-controller: deployment ready
► ghcr.io/fluxcd/helm-controller:v0.36.2
✔ kustomize-controller: deployment ready
► ghcr.io/fluxcd/kustomize-controller:v1.1.1
✔ notification-controller: deployment ready
► ghcr.io/fluxcd/notification-controller:v1.1.0
✔ source-controller: deployment ready
► ghcr.io/fluxcd/source-controller:v1.1.2
► checking crds
✔ alerts.notification.toolkit.fluxcd.io/v1beta2
✔ buckets.source.toolkit.fluxcd.io/v1beta2
✔ gitrepositories.source.toolkit.fluxcd.io/v1
✔ helmcharts.source.toolkit.fluxcd.io/v1beta2
✔ helmreleases.helm.toolkit.fluxcd.io/v2beta1
✔ helmrepositories.source.toolkit.fluxcd.io/v1beta2
✔ kustomizations.kustomize.toolkit.fluxcd.io/v1
✔ ocirepositories.source.toolkit.fluxcd.io/v1beta2
✔ providers.notification.toolkit.fluxcd.io/v1beta2
✔ receivers.notification.toolkit.fluxcd.io/v1
✔ all checks passed

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by kevinvalk** [comment]
So exact same behavior when running with v2.0.1 (server side)
```
► checking prerequisites
✗ flux 2.0.1 <2.1.2 (new version is available, please upgrade)
✔ Kubernetes 1.28.3+k3s1 >=1.24.0-0
► checking controllers
✔ helm-controller: deployment ready
► ghcr.io/fluxcd/helm-controller:v0.35.0
✔ kustomize-controller: deployment ready
► ghcr.io/fluxcd/kustomize-controller:v1.0.1
✔ notification-controller: deployment ready
► ghcr.io/fluxcd/notification-controller:v1.0.0
✔ source-controller: deployment ready
► ghcr.io/fluxcd/source-controller:v1.0.1
► checking crds
✔ alerts.notification.toolkit.fluxcd.io/v1beta2
✔ buckets.source.toolkit.fluxcd.io/v1beta2
✔ gitrepositories.source.toolkit.fluxcd.io/v1
✔ helmcharts.source.toolkit.fluxcd.io/v1beta2
✔ helmreleases.helm.toolkit.fluxcd.io/v2beta1
✔ helmrepositories.source.toolkit.fluxcd.io/v1beta2
✔ kustomizations.kustomize.toolkit.fluxcd.io/v1
✔ ocirepositories.source.toolkit.fluxcd.io/v1beta2
✔ providers.notification.toolkit.fluxcd.io/v1beta2
✔ receivers.notification.toolkit.fluxcd.io/v1
✔ all checks passed
```

Am I missing something obvious?

**Comment by stefanprodan** [comment]
Prune has no effect at force apply, it's only about GC behaviour. The issue is that we ignore any annotation when `force: true` https://github.com/fluxcd/pkg/blob/main/ssa/manager_apply.go#L346-L357

**Comment by stefanprodan** [comment]
No matter the annotations the Kubernetes API shouldn't return an immutable error if the PVC has no changes in Git. Can you post here `kubectl get pvc --show-managed-fields` please.

**Comment by kevinvalk** [comment]
This is after it is trying to be recreated by Flux (so termination state). I checked before deletion as well, only difference is the absence of `deletionGracePeriodSeconds` and `deletionTimestamp` fields.

The kustomize object also gives the message it has detected drift, but I have no clue how to see what the drift is. Maybe I could use flux diff on the server and just have a copy of the repo?

_Random thoughts, I am using Kustomize components, but none of these are touching the pvc object. I am also using substituteFrom which does touch the pvc object (`${storage_class_many}`) but that should be stable (I only write into into that variable on bootstrap). Maybe that has some "weird" interaction?_

```json
{
    "apiVersion": "v1",
    "kind": "PersistentVolumeClaim",
    "metadata": {
        "annotations": {
            "kustomize.toolkit.fluxcd.io/force": "disabled",
            "kustomize.toolkit.fluxcd.io/prune": "disabled",
            "kustomize.toolkit.fluxcd.io/ssa": "IfNotPresent",
            "pv.kubernetes.io/bind-completed": "yes",
            "pv.kubernetes.io/bound-by-controller": "yes",
            "volume.beta.kubernetes.io/storage-provisioner": "driver.longhorn.io",
            "volume.kubernetes.io/storage-provisioner": "driver.longhorn.io"
        },
        "creationTimestamp": "2023-11-09T08:18:37Z",
        "deletionGracePeriodSeconds": 0,
        "deletionTimestamp": "2023-11-09T08:19:38Z",
        "finalizers": [
            "kubernetes.io/pvc-protection"
        ],
        "labels": {
            "kustomize.component": "apps",
            "kustomize.toolkit.fluxcd.io/name": "apps",
            "kustomize.toolkit.fluxcd.io/namespace": "flux-system"
        },
        "managedFields": [
            {
                "apiVersion": "v1",
                "fieldsType": "FieldsV1",
                "fieldsV1": {
                    "f:metadata": {
                        "f:annotations": {
                            "f:kustomize.toolkit.fluxcd.io/force": {},
                            "f:kustomize.toolkit.fluxcd.io/prune": {},
                            "f:kustomize.toolkit.fluxcd.io/ssa": {}
                        },
                        "f:labels": {
                            "f:kustomize.component": {},
                            "f:kustomize.toolkit.fluxcd.io/name": {},
                            "f:kustomize.toolkit.fluxcd.io/namespace": {}
                        }
                    },
                    "f:spec": {
                        "f:accessModes": {},
                        "f:resources": {
                            "f:requests": {
                                "f:storage": {}
                            }
                        },
                        "f:storageClassName": {}
                    }
                },
                "manager": "kustomize-controller",
                "operation": "Apply",
                "time": "2023-11-09T08:18:37Z"
            },
            {
                "apiVersion": "v1",
                "fieldsType": "FieldsV1",
                "fieldsV1": {
                    "f:metadata": {
                        "f:annotations": {
                            "f:pv.kubernetes.io/bind-completed": {},
                            "f:pv.kubernetes.io/bound-by-controller": {},
                            "f:volume.beta.kubernetes.io/storage-provisioner": {},
                            "f:volume.kubernetes.io/storage-provisioner": {}
                        }
                    },
                    "f:spec": {
                        "f:volumeName": {}
                    }
                },
                "manager": "k3s",
                "operation": "Update",
                "time": "2023-11-09T08:18:39Z"
            },
            {
                "apiVersion": "v1",
                "fieldsType": "FieldsV1",
                "fieldsV1": {
                    "f:status": {
                        "f:accessModes": {},
                        "f:capacity": {
                            ".": {},
                            "f:storage": {}
                        },
                        "f:phase": {}
                    }
                },
                "manager": "k3s",
                "operation": "Update",
                "subresource": "status",
                "time": "2023-11-09T08:18:39Z"
            }
        ],
        "name": "repository-pvc",
        "namespace": "default",
        "resourceVersion": "1635436",
        "uid": "349b1bd1-69f9-4253-a858-25f12ec90f76"
    },
    "spec": {
        "accessModes": [
            "ReadWriteMany"
        ],
        "resources": {
            "requests": {
                "storage": "10Gi"
            }
        },
        "storageClassName": "longhorn",
        "volumeMode": "Filesystem",
        "volumeName": "pvc-349b1bd1-69f9-4253-a858-25f12ec90f76"
    },
    "status": {
        "accessModes": [
            "ReadWriteMany"
        ],
        "capacity": {
            "storage": "10Gi"
        },
        "phase": "Bound"
    }
}
```

# For completeness

**pvc-repositories.yaml**
```yaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: repository-pvc
  annotations:
    # We want to make sure we never throw this PVC away!
    kustomize.toolkit.fluxcd.io/prune: disabled
    kustomize.toolkit.fluxcd.io/force: disabled
    kustomize.toolkit.fluxcd.io/ssa: IfNotPresent
spec:
  accessModes:
    - ReadWriteMany
  resources:
    requests:
      storage: 10Gi
  storageClassName: ${storage_class_many}
```

**flux-apps.yaml**
```yaml
---
apiVersion: kustomize.toolkit.fluxcd.io/v1
kind: Kustomization
metadata:
  name: apps
  namespace: flux-system
spec:
  interval: 15m
  dependsOn:
    - name: infrastructure
  sourceRef:
    kind: GitRepository
    name: root
  path: ./apps
  prune: true
  force: true

  postBuild:
    substituteFrom:
      - kind: ConfigMap
        name: deployment-configuration
      - kind: Secret
        name: deployment-secrets
        optional: true
      - kind: ConfigMap
        name: cluster-configuration

  # NOTE: Important to be empty list because of components system
  components: []
```

**Comment by stefanprodan** [comment]
I strongly recommend to never set `force: true`, this is a global setting for when you apply a bunch of Job or other immutable objects, certainly not for infra and PVCs. I suggest you set `kustomize.toolkit.fluxcd.io/force: enabled` only on objects when you change immutable fields.

**Comment by kevinvalk** [comment]
Mmm interesting, I checked my git history and I was running with `force: true` on the flux kustomization for over a year without a problem. Another significant thing I changed recently (when this problem started) is I switched to K8s v1.28, maybe somehow that caused a difference in tracked/immutable fields? 

Regardless, thanks a lot! It did not immediately throw away the PVC anymore with the force only set on a few job objects. I will keep an eye on this for 24h and if it still does not happen I will close this issue.

Just a thought, maybe adding a warning to https://fluxcd.io/flux/components/kustomize/kustomizations/#force with your comment could help for future prosperity?
> I strongly recommend to never set force: true, this is a global setting for when you apply a bunch of Job or other immutable objects, certainly not for infra and PVCs. I suggest you set kustomize.toolkit.fluxcd.io/force: enabled only on objects when you change immutable fields.

Once again, thanks for quick investigation and all the work on Flux2!
